Unit Tests and Travis CI

djarotech opened the pull request:
In this commit I added unit tests for the exposed API endpoint methods. A number of them read from an output file that stores the expected output of the tests, tests.yml. Also, I added travis CI, so that it builds each time there is a new commit. Some helper methods and ordering was needed in provenance.go, like a String() change and an Ordered_map type that is useful for printing maps in a sorted order.

devdattakulkarni commented on README.md:
```diff
@@ -30,7 +30,7 @@ apt-get install -y gcc make socat git wget<br/>
 sudo tar -C /usr/local -xzf go1.10.3.linux-amd64.tar.gz <br/>
 export PATH=$PATH:/usr/local/go/bin <br/>
-export GOROOT=$PATH:/usr/local/go <br/>
+export GOROOT=/usr/local/go <br/>
```
Nice. This will fix
https://github.com/cloud-ark/kubeprovenance/issues/14

devdattakulkarni commented on pkg/provenance/provenance_test.go:
```diff
@@ -0,0 +1,521 @@
+	objLineage, newArgs := buildLineage()
+
+	users := updateUser(newArgs.Users, "daniel", "JEK873BUL!")
+	db := removeDatabase(newArgs.Databases, "logging")
```
Why removeDatabase in this test?

We can only keep this test for PasswordChange.

devdattakulkarni commented on pkg/provenance/provenance_test.go:
```diff
@@ -0,0 +1,521 @@
+	argMapTest["value1"] = "logging"
+
+	vOutput := objLineage.Bisect(argMapTest)
+	expected := "Version: 4"
```
Shouldn't the expected be "Version: 6" ?

devdattakulkarni replied:
I think I understand. Essentially we are saying that if an attribute is removed then Bisect will show the latest version in which that attribute existed. Correct?

djarotech replied:
Hmm good catch, however Bisect still finds the first version that attribute exists. Hence, removing logging in Version 6 doesn't really matter since it will still find Version 4, when it had the target config.

So this test is actually meaningless and I do not have the functionality right now to write a unit test for remove anyways. Shall I just remove this test? And the other similar test?

devdattakulkarni replied:
Ah I see. Yes, I agree that we should remove this and other such tests.

devdattakulkarni commented on pkg/provenance/provenance.go:
```diff
@@ -303,7 +365,7 @@ func (o ObjectLineage) Bisect(argMap map[string]string) string {
 	// it could find that in different parts of the spec. They both must be satisfied in the same map object
 	mapRelationships := buildAttributeRelationships(specs, allQueryPairs)
-	fmt.Printf("Query Attributes Map:%v\n", mapRelationships)
+	fmt.Printf("Query Attributes Same-parent-relationsips: %v\n", mapRelationships)
```
Same-parent-relationships.

devdattakulkarni commented:
@danielpygo Great work on this PR. The tests are comprehensive. 💯 
There are two small nits. They should be quick to fix. Otherwise this is ready to merge.

djarotech committed:
Unit Tests, Travis CI

In this commit I added unit tests for the exposed API endpoint methods. I added travis CI, so that it builds each time a new commit . And I have a small fix for minikube as well. Some helper methods and ordering was needed in provenance.go, like a String() change and an ordered_map type.

djarotech commented:
Ok, great .

devdattakulkarni commented:
Looks good!
